darshan-runtime/lib/darshan-core.c

```diff
@@ -470,13 +474,34 @@ void darshan_core_shutdown(int write_log)
     internal_timing_flag = final_core->config.internal_timing_flag;
 
 #ifdef __DARSHAN_ENABLE_MMAP_LOGS
-    /* remove the temporary mmap log files */
-    /* NOTE: this unlink is not immediate as it must wait for the mapping
-     * to no longer be referenced, which in our case happens when the
-     * executable exits. If the application terminates mid-shutdown, then
-     * there will be no mmap files and no final log file.
+    /* Flush memory-mapped data to the underlying file and then duplicate the
+     * mmap log file, in case an interrupt happens before the completion of
+     * this subroutine, leaving the log file corrupted. See github issue #1052.
      */
-    unlink(final_core->mmap_log_name);
+    msync(final_core->log_hdr_p, final_core->mmap_size, MS_SYNC);
+
+    /* Duplicate the log file, Below we on purpose ignore errors from open(),
+     * lseek(), read(), write(), and close(), as they are not fatal and should
+     * not affect the remaining tasks of this subroutine, i.e. log data file
+     * processing. Note if any of these system calls failed, then the contents
+     * of duplicated file will become useless, but such problem will be even
+     * more serious than being unable to duplicate the file.
+     */
+    int mmap_fd = open(final_core->mmap_log_name, O_RDONLY, 0644);
+    if (mmap_fd != -1) {
+        int dup_mmap_fd;
+        void *buf = (void*) malloc(final_core->mmap_size);
+        read(mmap_fd, buf, final_core->mmap_size);
+        close(mmap_fd);
+        snprintf(dup_log_fame, strlen(final_core->mmap_log_name), "%s.dup",
```

Comment:
This snprintf truncates the files name to something like `/tmp/carns_python3_id1445957_mmap-log-3779006568372861245-0.darsha` because the 2nd argument to is calling `strlen()` on the original `mmap_log_name`.  Maybe this size limit on the snprintf should be based on the size of the `dup_log_fname` instead?

Reply:
This should be fixed in 6a5c60f.

Reply:
Actually I have one other request here.  Sorry this didn't occur to me sooner.  I like that the duplicate file has "dup" in the file name, but it may be better if it weren't used as the file extension.  I find myself doing things like `ls /tmp/*.darshan` to look for intermediate files, when with this patch it would really need to be `ls /tmp/*.darshan /tmp/*.darshan.dup` to find them all.

Can we change the naming convention so that even the duplicate file names end with `.darshan`?

Reply:
Updated. Please give it a try.